SpatialGDK/Source/SpatialGDK/Private/Interop/RPCs/SpatialRPCService.cpp

```diff
@@ -0,0 +1,484 @@
+			if (EndpointComponentId == SpatialConstants::MULTICAST_RPCS_COMPONENT_ID)
+			{
+				RPCRingBufferUtils::MoveLastSentIdToInitiallyPresentCount(
+					Schema_GetComponentDataFields(*ComponentData),
+					RPCStore.LastSentRPCIds[EntityRPCType(EntityId, ERPCType::NetMulticast)]);
+			}
+
+			if (EndpointComponentId == SpatialConstants::CLIENT_ENDPOINT_COMPONENT_ID)
+			{
+				UE_LOG(LogSpatialRPCService, Error,
+					   TEXT("SpatialRPCService::GetRPCComponentsOnEntityCreation: Initial RPCs present on ClientEndpoint! EntityId: %lld"),
+					   EntityId);
+			}
+
+			Component.schema_type = *ComponentData;
+#if TRACE_LIB_ACTIVE
+			TraceKey Trace = InvalidTraceKey;
+			PendingTraces.RemoveAndCopyValue(EntityComponent, Trace);
+			Component.Trace = Trace;
+#endif
+			RPCStore.PendingRPCsOnEntityCreation.Remove(EntityComponent);
+		}
+		else
+		{
+			Component.schema_type = Schema_CreateComponentData();
+		}
+	}
+
+	return Components;
+}
+
+void SpatialRPCService::ProcessOrQueueIncomingRPC(const FUnrealObjectRef& InTargetObjectRef, SpatialGDK::RPCPayload InPayload)
+{
+	const TWeakObjectPtr<UObject> TargetObjectWeakPtr = NetDriver->PackageMap->GetObjectFromUnrealObjectRef(InTargetObjectRef);
+	if (!TargetObjectWeakPtr.IsValid())
+	{
+		UE_LOG(LogSpatialRPCService, Verbose, TEXT("The object has been deleted, dropping the RPC"));
+		return;
+	}
+
+	UObject* TargetObject = TargetObjectWeakPtr.Get();
+	const FClassInfo& ClassInfo = NetDriver->ClassInfoManager->GetOrCreateClassInfoByObject(TargetObject);
+
+	if (InPayload.Index >= static_cast<uint32>(ClassInfo.RPCs.Num()))
+	{
+		// This should only happen if there's a class layout disagreement between workers, which would indicate incompatible binaries.
+		UE_LOG(LogSpatialRPCService, Error, TEXT("Invalid RPC index (%d) received on %s, dropping the RPC"), InPayload.Index,
+			   *TargetObject->GetPathName());
+		return;
+	}
+	UFunction* Function = ClassInfo.RPCs[InPayload.Index];
+	if (Function == nullptr)
+	{
+		UE_LOG(LogSpatialRPCService, Error, TEXT("Missing function info received on %s, dropping the RPC"), *TargetObject->GetPathName());
+		return;
+	}
+
+	const FRPCInfo& RPCInfo = NetDriver->ClassInfoManager->GetRPCInfo(TargetObject, Function);
+	const ERPCType Type = RPCInfo.Type;
+
+	IncomingRPCs.ProcessOrQueueRPC(InTargetObjectRef, Type, MoveTemp(InPayload));
+}
+
+void SpatialRPCService::ClearPendingRPCs(Worker_EntityId EntityId)
+{
+	IncomingRPCs.DropForEntity(EntityId);
+}
+
+void SpatialRPCService::PeriodicallyProcessIncomingRPCs()
+{
+	FTimerHandle IncomingRPCsPeriodicProcessTimer;
+	TimerManager->SetTimer(
+		IncomingRPCsPeriodicProcessTimer,
+		[this]() {
+			IncomingRPCs.ProcessRPCs();
+		},
+		GetDefault<USpatialGDKSettings>()->QueuedIncomingRPCRetryTime, true);
+}
+
+EPushRPCResult SpatialRPCService::PushRPCInternal(const Worker_EntityId EntityId, const ERPCType Type, RPCPayload&& Payload,
```

Comment:
not a big fan of the name as both PushRPC and PushRPCInternal are kind of internal.

Reply:
This was named because PushRPC is public, and PushRPCInternal is private

Reply:
@jessicafalk I agree, and this isn't the only time this comes up. This is one of the functions I've just moved around in an effort to touch as little as possible. I can try and work out better names for these if you feel strongly- I imagine these will be changed again in the not too distant future once it becomes easier to now they're all split.